File: 84.LargestRectangleinHistogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [] # (index, height)
        N = len(heights)
        max_height = 0

        for i in range(len(heights)):
            if not stack:
                stack.append((i, heights[i]))
                continue

            if stack[-1][1] < heights[i]:
                stack.append((i, heights[i]))
                continue

            if stack[-1][1] > heights[i]:            
                while stack and stack[-1][1] > heights[i]:
                    idx = stack[-1][0]
                    max_height = max(max_height, (i - stack[-1][0]) * stack[-1][1])
                    stack.pop()
                stack.append((idx, heights[i]))

        while stack:
            idx, height = stack.pop()
            max_height = max(max_height, (N - idx) * height)

        return max_height

File: 84.LargestRectangleinHistogram.java
class Solution {
    class Pair{
        int idx, height;
        public Pair(int idx, int height){
            this.idx = idx;
            this.height = height;
        }
    }
    public int largestRectangleArea(int[] heights) {
        Stack<Pair> stack = new Stack<>();
        int N = heights.length;
        int maxArea = 0;

        for(int i = 0; i < N; i++){
            if(stack.isEmpty() || stack.peek().height < heights[i]){
                stack.add(new Pair(i, heights[i]));
                continue;
            }
            else{
                int index = i;
                while(!stack.isEmpty() && stack.peek().height >= heights[i]){
                    Pair p = stack.pop();
                    index = p.idx;
                    maxArea = Math.max(maxArea, (i - p.idx) * p.height);
                }
                stack.add(new Pair(index, heights[i]));
            }
        }
        while(!stack.isEmpty()){
            Pair s = stack.pop();
            maxArea = Math.max(maxArea, (N - s.idx) * s.height);
        }
        return maxArea;

    }
}
